simulation/cdm_data.py

```diff
@@ -0,0 +1,206 @@
+            self._observation_period_observation_period_start_date.collect(),
+            self._observation_period_observation_period_end_date.collect()],
+            names=["person_id", "observation_period_id", "observation_period_start_date",
+                   "observation_period_end_date"])
+        file_name = "part{:04d}.parquet".format(partition_i + 1)
+        _create_folder_if_not_exists(os.path.join(root_folder, "person"))
+        pq.write_table(person, os.path.join(root_folder, "person", file_name))
+        _create_folder_if_not_exists(os.path.join(root_folder, "visit_occurrence"))
+        pq.write_table(visit_occurrence, os.path.join(root_folder, "visit_occurrence", file_name))
+        _create_folder_if_not_exists(os.path.join(root_folder, "condition_occurrence"))
+        pq.write_table(condition_occurrence, os.path.join(root_folder, "condition_occurrence", file_name))
+        _create_folder_if_not_exists(os.path.join(root_folder, "observation_period"))
+        pq.write_table(observation_period, os.path.join(root_folder, "observation_period", file_name))
+
+    def log_statistics(self, partition_i: int):
+        logging.debug("Partition %s persons: %s", partition_i, len(self._person_person_id))
+        logging.debug("Partition %s visit_occurrences: %s", partition_i, len(self._visit_occurrence_person_id))
+        total_days = np.sum(np.subtract(self._observation_period_observation_period_end_date.collect(),
+                                        self._observation_period_observation_period_start_date.collect()).view(
+            np.int64))
+        logging.debug("Partition %s mean days between visits: %.1f", partition_i,
+                      total_days / len(self._visit_occurrence_person_id))
+        logging.debug("Partition %s mean concepts per visit: %.1f", partition_i,
+                      len(self._condition_occurrence_person_id) / len(self._visit_occurrence_person_id))
+        logging.debug("Partition %s unique concept count: %s", partition_i,
+                      len(np.unique(self._condition_occurrence_condition_concept_id.collect())))
+        logging.debug("Partition %s percent ER visits: %.1f%%", partition_i,
+                      100 * sum(self._visit_occurrence_visit_concept_id.collect() == 9203) /
+                      len(self._visit_occurrence_person_id))
+
+
+class _Labels:
+
+    person_id: DynamicArray
+    label: DynamicArray
+
+    def __init__(self):
+        super().__init__()
+        self.person_id = DynamicArray(np.int64)
+        self.label = DynamicArray(bool)
+
+
+class CdmDataWithLabels(CdmData):
+
+    def __init__(self):
+        super().__init__()
+        self._concept_id_to_labels: defaultdict[int, _Labels] = defaultdict(_Labels)
+
+    def add_label(self, person_id: int, concept_id: int, label: bool):
+        labels = self._concept_id_to_labels[concept_id]
+        labels.person_id.append(person_id)
+        labels.label.append(label)
```

Comment:
It looks like we store a `person_id` and `label` pair for every `concept_id`, e.g. there are 100 folders generated containing the same `person_id` and `label` pairs after running the simulation, could you elaborate a bit more on how this data will be used for downstream prediction?

Reply:
Each folder is a partitioned Parquet table corresponding to a concept ID. in each table the label for a person ID will be True if the concept ID was observed in the prediction window, and False if it was not. So we can use this to train and validate a prediction model (and one way to train a prediction model is to fine-tune a pre-trained model)